projects/graph/graph.py

```diff
@@ -11,45 +11,142 @@ def add_vertex(self, vertex):
+        while q.size() > 0:
+            # Dequeue the first vertex
+            v = q.dequeue()
+            # If that vertex has not been visited...
+            if v not in visited:
+                # Mark it as visited
+                visited.add(v)
+                print(f'bft print {v}')
+                # Then add all of its neighbors to the back of the queue
+                for neighbor in self.vertices[v]:
+                    q.enqueue(neighbor)
+
     def dft(self, starting_vertex):
         """
         Print each vertex in depth-first order
         beginning from starting_vertex.
         """
-        pass  # TODO
+        visited = set()
+        s = Stack()
+        s.push(starting_vertex)
+        while s.size() > 0 :
+            v = s.pop()
+            if v not in visited:
+                visited.add(v)
+                print(f'dft print {v}')
+                for neighbor in self.vertices[v]:
+                    s.push(neighbor)
+
     def dft_recursive(self, starting_vertex):
         """
         Print each vertex in depth-first order
         beginning from starting_vertex.
         This should be done using recursion.
         """
-        pass  # TODO
+        visited = set()
+        def inner_recursive(vertex):
+            if vertex in visited:
+                print(f' never print because starting vertex is not visited but immediately added.')
+                return
+            print(f'DFS recursive vertex {vertex}')
+            visited.add(vertex)
+            for neighbor in self.vertices[vertex]:
+                if neighbor not in visited:
+                    inner_recursive(neighbor)
+
+        # print(f'dft recursive')
+        # visited = set()
+        # def inner_recursive(vertex):
+        #     visited.add(vertex)
+        #     
+        #     print(f'Recursive vertex = {vertex} {visited}')
+        #     for neighbor in self.vertices[vertex]:
+        #         if neighbor not in visited:
+        #             inner_recursive(neighbor)
+
+        #     print(f' Nothing printed from {visited} because this visited is not part of recursive loop.')
+  
+        inner_recursive(starting_vertex)
+
     def bfs(self, starting_vertex, destination_vertex):
         """
         Return a list containing the shortest path from
         starting_vertex to destination_vertex in
         breath-first order.
         """
-        pass  # TODO
+        # Create an empty set to store visited nodes
+        visited = set()
+        # Create an empty Queue and enqueue A PATH TO the starting vertex
+        q = Queue()
+        path = [starting_vertex]
+        q.enqueue(path)
+        # While the queue is not empty...
+        while q.size() > 0 :
+            # Dequeue the first PATH
+            path = q.dequeue()
+            # GRAB THE VERTEX FROM THE END OF THE PATH
+            v = path[-1]
+            # IF VERTEX = TARGET, RETURN PATH
+            if v == destination_vertex :
```

Comment:
does it work? would think that you would need to have the check if v is == to destination nested inside the not in visited if statement